src/ccontrol/MergingHandler.cc

```diff
@@ -38,22 +43,269 @@
+                                                " exceeds the hard limit of " +
+                                                to_string(ProtoHeaderWrap::PROTOBUFFER_HARD_LIMIT) +
+                                                ", file: " + httpUrl);
+                        }
+                        // Extend the message buffer (if needed). Note that buffer never gets
+                        // truncated to avoid excessive memory deallocations/allocations.
+                        if (msgBufSize < msgSizeBytes) {
+                            msgBufSize = msgSizeBytes;
+                            msgBuf.reset(new char[msgBufSize]);
+                        }
+                    }
+                } else {
+                    // Continue or finish reading the message body.
+                    size_t const bytes2read =
+                            std::min((size_t)msgSizeBytes - msgBufNext, (size_t)(end - next));
+                    std::memcpy(msgBuf.get() + msgBufNext, next, bytes2read);
+                    next += bytes2read;
+                    offset += bytes2read;
+                    msgBufNext += bytes2read;
+                    if (msgBufNext == msgSizeBytes) {
+                        // Done reading message body.
+                        msgBufNext = 0;
+                        // Parse and evaluate the message.
+                        bool const success = messageIsReady(msgBuf.get(), msgSizeBytes);
+                        if (!success) {
+                            throw runtime_error(context + "message processing failed at offset " +
+                                                to_string(offset - msgSizeBytes) + ", file: " + httpUrl);
+                        }
+                        // Reset the variable to prepare for reading the next header & message (if any).
+                        msgSizeBytes = 0;
+                    }
+                }
+            }
+        });
+        if (msgSizeBufNext != 0) {
+            throw runtime_error(context + "short read of the message header at offset " +
+                                to_string(offset - msgSizeBytes) + ", file: " + httpUrl);
+        }
+        if (msgBufNext != 0) {
+            throw runtime_error(context + "short read of the message body at offset " +
+                                to_string(offset - msgSizeBytes) + ", file: " + httpUrl);
+        }
+    } catch (exception const& ex) {
+        LOGS(_log, LOG_LVL_ERROR, ex.what());
+        success = false;
+    }
+
+    // Remove the file from the worker if it still exists. Report and ignore errors.
+    // The files will be garbage-collected by workers.
+    try {
+        HttpClient remover("DELETE", httpUrl);
+        remover.read([](char const* inBuf, size_t inBufSize) {});
+    } catch (exception const& ex) {
+        LOGS(_log, LOG_LVL_WARN, context << "failed to remove " << httpUrl << ", ex: " << ex.what());
```

Comment:
The intent here is to avoid crashing Qserv `Czar` for any reason. All exceptions are caught and reported via LSST Logger. After that (catching an exception) the algorithm returns `false` to the calling code. This decision was made intentionally to make the error reporting of this algorithm compatible with the one of context from which the algorithm is being called:
```
bool MergingHandler::flush(...)
```
Note that `flush` is expected to report `false` for any problems encountered while attempting to merge results.
Also, I'm not sure what would happen if an exception would be thrown upstream, and what else would be able to identify and correctly interpret such exceptions in the upstream call chain.
Another advantage of catching all exceptions in the algorithm is that this would allow to properly report the location of the reported problems closer to the source of those.